### ISSUE TITLE
[HOLD for payment 2023-09-20] [$1250] Cache Pods to speed up iOS deploys

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#18623]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

### Problem

Our slowest platform to deploy, by far, is iOS. Part of the reason for that is because we do a fresh install of all our pods on every deploy, which can take [up to 12 minutes](https://github.com/Expensify/App/actions/runs/4921036845/jobs/8790467358). This is a problem because it slows our release cycle, particularly when we want to cherry-pick fixes to our staging environment.

### Solution

Use https://github.com/actions/cache to cache our cocoapods and hopefully speed up the iOS deploy. Even a 5 minute improvement would pay dividends in the long run.

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010c613f2c1b3960d0</li>
        <li>Upwork Job ID: 1655741689607213056</li>
        <li>Last Price Increase: 2023-09-10</li>
    </ul>
</details>

**Comment by roryabraham** [comment]
I'm not sure why we got both @zanyrenney _and_ @greg-schroeder assigned here. Going to give it to unassign @greg-schroeder since @zanyrenney was assigned first.

**Comment by nbhargava** [comment]
## Proposal
We're going to use the `actions/cache` action as mentioned in the original ticket and hash the existing Podfile in order to determine whether we can re-use the existing cache or whether we have to do a fresh install. We'll update `App/.github/workflows/testBuild.yml` to include something that looks like:
```
jobs:
  [...]
  iOS:
    steps:
    - [...]
    - uses: actions/cache@v2
      with:
        path: ios/Pods
        key: ${{ runner.os }}-pods-${{ hashFiles('**/Podfile.lock') }}
        restore-keys: |
          ${{ runner.os }}-pods-
    - name: Install cocoapods
      [...]
```

It may require a bit of fiddling to get the exact invocation right, but this should ensure that we have a cached set of Pods that we can recreate from when possible.

We should be able to test this pretty easily within the context of a single PR by issuing two subsequent builds. We should see a marked speed up in the second build after we issue the first one due to a cache hit.

**Comment by nbhargava** [comment]
Contributor details
Your Expensify account email: expensify@nikhil.dev
Upwork Profile Link: https://www.upwork.com/freelancers/~017eb2ad704f8a883c

**Comment by mananjadhav** [comment]
I think @nbhargava's proposal [here](https://github.com/Expensify/App/issues/18623#issuecomment-1539272647) is good enough.

@roryabraham All yours.

**Comment by nbhargava** [comment]
Just submitted on Upwork -- hoping to get this in by Friday EOD.

**Comment by nbhargava** [comment]
Hi all -- I still have not been granted the job on Upwork. I would love to get this in within the 3 business day timeline to be eligible for the bonus (and still believe I can do so) but am not going to continue working on this until I get some clarity in here in case I'm misunderstanding the process.

Thanks!

**Comment by roryabraham** [comment]
@nbhargava I've extended the offer on Upwork (hopefully I did I right, I don't usually deal with the Upwork piece myself). Rest assured we will honor the payment for this issue. Please reopen the pull request and get it ready for review 👍🏼

**Comment by nbhargava** [comment]
Sounds great -- PR should be good to go!

**Comment by mananjadhav** [comment]
Thanks @nbhargava. Good to see your first PR this fast. I'll review this in a few hours. 

Meanwhile I would recommend you can go through the [contributing guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) to understand our job assignment and payout process.

**Comment by roryabraham** [comment]
If I'm not mistaken @nbhargava [has contributed before](https://github.com/Expensify/App/pull/1318), and did a great job. Welcome back @nbhargava 🙂

It was a couple years ago when our open-source program was a lot less mature than it is today, so it's a good idea to read the up-to-date contributing guidelines as @mananjadhav has suggested.

**Comment by roryabraham** [comment]
Yeah, @zanyrenney this one is a bit different than a typical PR because GitHub Actions immediately "deploy to prod". We can see that in the event of a cache hit such as [here](https://github.com/Expensify/App/actions/runs/5024004782/jobs/9009247872), the cocoapods install is down to ~2 mins instead of ~12 mins. So I think we can pay and close this out right away.

**Comment by zanyrenney** [comment]
Hey @roryabraham did you close the job posting? I also can't see any of the hires: 
<img width="967" alt="2023-05-24_14-21-36" src="https://github.com/Expensify/App/assets/56830058/529bebc2-a97a-4af8-b2aa-86372e1b8fea">

Can you confirm that the **two** hires should be [mananjadhav](https://github.com/mananjadhav) and [nbhargava](https://github.com/nbhargava) and that the posting should be open?

I can recreate it if the post is closed accidentally?

**Comment by mananjadhav** [comment]
@zanyrenney Yeah it should be me and @nbhargava

**Comment by zanyrenney** [comment]
Going to recreate a job because the one @roryabraham made it isn't letting me invite you !

**Comment by zanyrenney** [comment]
Recreated job post and invited you both  @nbhargava @mananjadhav https://www.upwork.com/jobs/~012efa572c36bd3bfb

**Comment by zanyrenney** [comment]
That's so odd 😂 i changed the title when recreating the job to the correct title, have changed it again!

**Comment by mananjadhav** [comment]
Thanks, I can see the updated title now. Accepted 🤠

**Comment by zanyrenney** [comment]
still waiting on @nbhargava to accept, payout is due!

**Comment by zanyrenney** [comment]
Going OOO and this is pretty active so needs management by bug0 team member, as per the [new process ](https://expensify.slack.com/archives/CK49UADB9/p1684182291511799)to stay assigned but assign another team member, please can you keep a watch of this. I am OOO until 8th June. Thanks!

**Comment by mananjadhav** [comment]
@zanyrenney @michaelhaxhiu I've accepted the offer, but payout is pending.

**Comment by michaelhaxhiu** [comment]
On it.

**Comment by michaelhaxhiu** [comment]
So there is no speed bonus for this job just to be clear?

**Comment by mananjadhav** [comment]
> So there is no speed bonus for this job just to be clear?

I think there would be.

**Comment by michaelhaxhiu** [comment]
All done here.

**Comment by mananjadhav** [comment]
@roryabraham any specific reason we reverted the PR?

**Comment by roryabraham** [comment]
Yeah, we kept getting this error `Sandbox is not in sync with Podfile.lock`

**Comment by michaelhaxhiu** [comment]
Un-assigning myself since original BZ is back from OOO

**Comment by mananjadhav** [comment]
@zanyrenney Do we want to add the `Help Wanted` label here?

**Comment by mananjadhav** [comment]
@zanyrenney @michaelhaxhiu if anyone of you can help with the previous comment?

**Comment by roryabraham** [comment]
Let's add it, but we should hold a very very high standard for testing, including an explanation for why pod installs were failing when we had this implemented previously.

**Comment by myrs** [comment]
Hi! I have a proposal, but it looks like [job posting](https://www.upwork.com/jobs/~012efa572c36bd3bfb) on Upwork is expired. I would be glad to take it once it's available. Thank you!

**Comment by myrs** [comment]
Contributor details
Your Expensify account email: mihail.nso@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0109215bee1a2d9638

**Comment by roryabraham** [comment]
@zanyrenney can you please update the Upwork job posting here so @myrs feels confident to post their proposal?

FWIW @myrs I can assure you that if your proposal is accepted and you complete the PR we'll honor the payment listed to this issue in accordance with our [Contributing Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md).

**Comment by myrs** [comment]
@roryabraham got you! I'm leaving the proposal below. Happy to take the tasks if it sounds good to you and open to further discussion.

## Explanation of the error you are experiencing
The error `Sandbox is not in sync with Podfile.lock` you have experienced with the original solution happens when `ios/Podfile.lock` and `ios/Pods/Manifest.lock` are not in sync. 

This is a safety check that Xcode performs during the build to ensure the lock version of the project (`ios/Podfile.lock`) is the same as the lock version on the local machine (`ios/Pods/Manifest.lock`). You can find a more detailed discussion [here](https://stackoverflow.com/a/65783259/4267649).

From my experience, this can happen for two reasons described below (but there can be more to it).
**First**. Different versions of `cocoapods` are used to install `pods` when running `pod install` on different machines. We can imagine the following scenario: 
- Imagine a developer has `cocoapods = 1.12.0` and the CI/CD pipeline machine  `cocoapods = 1.12.1`
- A developer has run `pod install` on their machine. Las line of `ios/Podfile.lock` is `COCOAPODS: 1.12.0`
- First CI/CD pipeline run. The cache is not found: it runs `pod install`. This updates `ios/Podfile.lock` and `ios/Pods/Manifest.lock` to have `COCOAPODS: 1.12.1` as the last line
- Second CI/CD pipeline run. The matching cache is found. Importantly the `ios/Podfile.lock` still contains `COCOAPODS: 1.12.0` (from the repository). But the cache contains the `ios/Pods/Manifest.lock` with `COCOAPODS: 1.12.1`. Hence, the build phase fails with `Sandbox is not in sync with Podfile.lock`.

**Second**. `Podfile.lock` is out of sync with `package-lock.json`. We can imagine a situation when a developer run `npm install` updating `package-lock.json`, but never run `pod install` and made a commit. The scenario develops as follows:
- First CI/CD pipeline run. The cache is not found: it runs `pod install`. This updates `ios/Podfile.lock` and `ios/Pods/Manifest.lock` with new dependency versions, e.g. `Expo (46.0.0) -> Expo (46.0.1)`. 
- Second CI/CD pipeline run. The matching cache is found. But this cache is found for the version of `ios/Podfile.lock` form repository, where `Expo (46.0.0)`. As  `ios/Pods/Manifest.lock` contains `Expo (46.0.1)`, the build fails with error: `Sandbox is not in sync with Podfile.lock`.


### Proposal
My proposal mitigates above reasons and, most importantly, prevents the build from failing with the `Sandbox is not in sync with Podfile.lock` for any other possible reasons of `ios/Podfile.lock` and `ios/Pods/Manifest.lock` mismatch.

To achieve this, we will run `pod install` again if `ios/Podfile.lock` and `ios/Pods/Manifest.lock` don't match even when the cache is found. To archive this, we will add something like this in the pipeline:

```
jobs:
  [...]
  iOS:
    steps:
    - [...]
    
    - name: cache Pod dependencies
      uses: actions/cache@v3
      id: pods-cache
      with:
        path: ios/Pods
        key: ${{ runner.os }}-pods-${{ hashFiles('**/Podfile.lock') }}

    - name: compare Podfile.lock and Manifest.lock
      id: compare-pods-lock-files
      run: |
          echo "are_lock_files_same=${{ hashFiles('ios/Podfile.lock') == hashFiles('ios/Pods/Manifest.lock') }}" >> $GITHUB_OUTPUT

    - name: Install cocoapods
      if:  ${{ (steps.pods-cache.outputs.cache-hit != 'true') || (!steps.compare-pods-lock-files.are_lock_files_same) }}
      working-directory: ./ios
      run: |
          bundle update && bundle exec pod install
```

**Comment by mananjadhav** [comment]
Thanks for the proposal @myrs. I understand your solution of comparing the files and rerunning the `pod install`. I still grasp what's the root cause of the previous solution? I am not sure how developer pod install would affect CD/CI pipeline?  Caching is supposed to be only GH actions only?

Also I feel comparing the file and rerunning the pod install isn't a workaround than fixing the root cause.

**Comment by myrs** [comment]
Hi @mananjadhav, happy to address your questions!

### TLDR;
The root cause of the message `Sandbox is not in sync with Podfile.lock` is any discrepancy between `ios/Podfile.lock` and `ios/Pods/Manifest.lock` when running the `fastane` command (`xcodebuild` command under the hood). While the root cause can be tricky to tackle 100%, the proposal ensures that the build will never fail with this error and that there will be considerable speed up even when the "lock" files are not in sync.

### Addressing explanation
Maybe the idea to include a "developer" in my original answer was an unnecessary complication 😉 I will try again without it. 

**Summary**. The caching indeed happens on GH actions only. Sometimes running `pod install` inside GH _can_ change the `Podfile.lock` and the `Pods/Manifest.lock`. But as only `Pods/Manifest.lock` is cached, next run, the cached `Pods/Manifest.lock` will differ from the `Podfile.lock` (from repo). This difference leads to the error `Sandbox is not in sync with Podfile.lock`

**First run**. Cach miss. `pod install` command is run and [for some reason, see my original answer] `Podfile.lock` and `Pods/Manifest.lock` are updated. Here, the most important bit💡: the updated `Pods/Manifest.lock` _is_ stored it cache. _But_ the updated `Podfile.lock` is not stored anywhere (i.e. it's not committed back to repo)

**Second run**. Cache found. This cache contains the updated `Pods/Manifest.lock` from the previous run. But, as we can see from the previous step, the `Podfile.lock` is an [old] file from the repository. So during the build we get an error `Sandbox is not in sync with Podfile.lock`.

### Addressing the solution being a workaround
I agree this solution is a workaround in a sense it does not prevent `Podfile.lock` from being out of sync with `Pods/Manifest.lock`. At the same time, it can't be 100% prevented. There are several additional things to try, but I feel they are out of the scope of this task and need either implementation of best practices / additional automation / updates to GH pipelines, e.g.:
- Ensure developers always run `bundle exec pod install` after `npm intall` and before committing their code
- Ensure developers run `bundle exec pod install` and not `pod install`. Running later with a local cocoapods version different than in `Podfile.lock`, will result in changes in `Podfile.lock`. In this case, the first run in GH will result in a cache miss, but in the second run will result in the `Sandbox is not in sync with Podfile.lock` error.
- After GH action runs, commit`Podfile.lock`

This being said, I feel the proposal addresses speeding up goal and fixes the original solution:
- It guarantees the build will never fail
- It will speed up the build when cache is present and `Podfile.lock` and `Pods/Manifest.lock` are in sync
- Importantly, it speeds up the build even when we have cache, but `Podfile.lock` and `Pods/Manifest.lock` are not in sync. In this case, `bundle update && bundle exec pod install` command will still be run (as `steps.compare-pods-lock-files.are_lock_files_same != true`) _but_ most of the pods will be present in the cache. It means it runs much quicker: in my local tests, the command runs in under 2 minutes (vs 10+ minutes without any cache).

### Proposal correction
In the proposal I've shared above, the condition should be changed to:
```
if:  steps.pods-cache.outputs.cache-hit != 'true' || steps.compare-pods-lock-files.outputs.are_lock_files_same != 'true'
```

I hope this brings more clarity to my proposal, and happy to address any additional questions.

**Comment by zanyrenney** [comment]
@roryabraham please could you chime in on the approach above to see if it's the right direction, thanks!

**Comment by mananjadhav** [comment]
@roryabraham quick bump? and should we ask someone from the agency to look into this?

**Comment by myrs** [comment]
Hi, @roryabraham, Thank you for the feedback and happy to take the job! I just checked the [job posting](https://www.upwork.com/jobs/~010c613f2c1b3960d0) and it's still marked as "no longer available". Could you or @mananjadhav have a look?

**Comment by mananjadhav** [comment]
@myrs, @zanyrenney Can help you here. Once you're assigned on GH, it is as good as being hired on Upwork. You can definitely wait for getting hired on Upwork, but rest assured Expensify will honor the contract once assigned on Github.

**Comment by roryabraham** [comment]
Yes, we can will get the job posting back up but @myrs rest assured that we'll honor the $1000 payment for this issue once the solution has been deployed for 7 days 🙂

**Comment by myrs** [comment]
Perfect! Thank you, I will hand in the PR this Friday.

**Comment by myrs** [comment]
Hi @roryabraham and @mananjadhav, I've created the [PR](https://github.com/Expensify/App/pull/27102) 🙂 

I have also included the update to `verifyPodfile.sh` as suggested by @roryabraham [here](https://github.com/Expensify/App/issues/18623#issuecomment-1705427286). If I understand correctly, including this update the bounty would be `$1250`.

Looking forward to your review!

**Comment by roryabraham** [comment]
PR was merged without C+ review, removing @mananjadhav. Thanks everyone!

**Comment by myrs** [comment]
@zanyrenney could you please help me with [Upwork job](https://www.upwork.com/jobs/~010c613f2c1b3960d0), it still appears with the status "This job is no longer available". As far as I understand, I need to get assigned to this job in order to receive the payment. Thank you!

**Comment by myrs** [comment]
@zanyrenney @roryabraham hi! Do you still need any information/actions for me to complete the payment? Thank you!

**Comment by zanyrenney** [comment]
for some reason the job is closed, i'll make another.

**Comment by zanyrenney** [comment]
Can't find you on upwork under

Misha Yaroshevskiy or Myrs. @myrs can you share your details so i can payout?

**Comment by zanyrenney** [comment]
https://www.upwork.com/jobs/~01d0e831f76674bb8f

**Comment by myrs** [comment]
Hi @zanyrenney , just submitted my application to the job

**Comment by zanyrenney** [comment]
Offer sent, please accept for payout @myrs

**Comment by zanyrenney** [comment]
Payment summary:

- contributor @myrs - $1250
- contributor plus - merged withuot C+ review
- issue reporter - N/A

**Comment by zanyrenney** [comment]
paid out and contract ended. thanks!
